### PR TITLE
Adding auth

### PR DESCRIPTION
**Pull request by SlyPuffin**
Auth is now functional and I've got the status and questions re-written below:

Resources:
- https://github.com/vercel/next.js/tree/canary/examples/auth0
- https://auth0.com/blog/ultimate-guide-nextjs-authentication-auth0/

Status:
Auth works with logging in and out with the NavBar. Pages display differently depending on login status.

Discussion Points:
- RESOLVED The example uses hooks, which won't work in component classes. Navbar is a component class which means 1) I use auth without hooks, or 2) We change Navbar to a function.
> I switched up the Navbar code to be a function which helped get everything working.
- The flow I have in mind is as follows, would be happy to discuss:

> - RESOLVED Landing page
>> Not authorized -> Short description of application and call to log-in
>> Authorized -> Welcome, [name]. & Profile + info (user info, deck count, dates studied, etc)
> - RESOLVED  Deck page
>> Not authorized -> "Log in to make your own decks and start studying"-type thing
>> Authorized -> Deck list
> - NO CHANGE Edit page (accessed from within deck list)
>> Not authorized -> Re-direct to deck page
>> Authorized -> Editing the deck page
> - NO CHANGE Practice page (accessed from within deck list)
>> Not authorized -> Re-direct to deck page
>> Authorized -> Review the deck

NEW Q: This does not connect to the database yet. Auth leaves the app and uses Auth0s page at the moment, so there needs to be a step to connect new users to the database.
At first I tried google auth, but this does not return the user's e-mail address and I restricted the app to Auth0, but this still makes it a little tricky to figure out at what event the user should be added to the database. Looking forward to chatting about this next time!

### CODE REVIEW

File: src/auth/user.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
+import { useState, useEffect } from "react";
+
+export async function fetchUser(cookie = "") {
+  if (typeof window !== "undefined" && window.__user) {
```
Comment (nleech16) [inline src/auth/user.js:4]: When will window be undefined?

File: src/auth/user.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    () => {
+      if (!loading && user) {
+        return;
+      }
+      setLoading(true);
+      let isMounted = true;
+
+      fetchUser().then((user) => {
+        console.log("user!");
+        console.log(user);
```
Comment (nleech16) [inline src/auth/user.js:53]: Probs don't want to check this in?

File: src/pages/api/callback.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import auth0 from "@/auth/auth0";
+
+export default async function callback(req, res) {
```
Comment (nleech16) [inline src/pages/api/callback.js:3]: maybe should have more specific name referencing auth? (or could it be used for other purposes?)

File: src/pages/decks/index.tsx [hunk 4]
```diff
@@ -1,12 +1,16 @@
...
-import { Deck } from "@prisma/client";
+import { useFetchUser } from "@/auth/user";
 
 export default function Home() {
-  const userQuery = trpc.useQuery(["get-user"], {
-    refetchOnWindowFocus: false,
-  });
+  const { user, loading } = useFetchUser();
+  const userQuery = trpc.useQuery(
+    ["get-user", { email: user?.name ? user?.name : "" }],
```
Comment (nleech16) [inline src/pages/decks/index.tsx:9]: might be worth a comment about assuming that user.name is email

File: src/pages/decks/index.tsx [hunk 5]
```diff
@@ -54,25 +58,43 @@ export default function Home() {
     }
   }
 
-  if (isLoading) {
+  if (loading || (user && isLoading)) {
```
Comment (nleech16) [inline src/pages/decks/index.tsx:61]: isUserLoading and areDecksLoading maybe?

Comment (SlyPuffin) [reply src/pages/decks/index.tsx:61]: Agreed! This was a too-lazy-to-refactor situation. I'll get this cleaned up :)

**Pull request merged**